src/Passes/Source/Passes/TransformationRule/TransformationRule.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Passes/TransformationRule/TransformationRule.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <fstream>
+#include <iostream>
+
+namespace microsoft
+{
+namespace quantum
+{
+    TransformationRulePass::TransformationRulePass(RuleSet&& rule_set)
+      : rule_set_{std::move(rule_set)}
+    {
+    }
+
+    llvm::PreservedAnalyses TransformationRulePass::run(
+        llvm::Function& function,
+        llvm::FunctionAnalysisManager& /*fam*/)
+    {
+        replacements_.clear();
+
+        // For every instruction in every block, we attempt a match
+        // and replace.
+        for (auto& basic_block : function)
+        {
+            for (auto& instr : basic_block)
+            {
+                rule_set_.matchAndReplace(&instr, replacements_);
+            }
+        }
+
+        // Applying all replacements
+        for (auto it = replacements_.rbegin(); it != replacements_.rend(); ++it)
+        {
+            auto instr1 = llvm::dyn_cast<llvm::Instruction>(it->first);
+            if (instr1 == nullptr)
+            {
+                llvm::errs() << "; WARNING: cannot deal with non-instruction replacements\n";
+                continue;
+            }
+
+            // Cheking if have a replacement for the instruction
+            if (it->second != nullptr)
+            {
+                // ... if so, we just replace it,
+                auto instr2 = llvm::dyn_cast<llvm::Instruction>(it->second);
+                if (instr2 == nullptr)
+                {
+                    llvm::errs() << "; WARNING: cannot replace instruction with non-instruction\n";
+                    continue;
+                }
+                llvm::ReplaceInstWithInst(instr1, instr2);
+            }
+            else
+            {
+                // ... otherwise we delete the the instruction
+                // Removing all uses
+                if (!instr1->use_empty())
+                {
+                    instr1->replaceAllUsesWith(llvm::UndefValue::get(instr1->getType()));
+                }
+
+                // And finally we delete the instruction
+                instr1->eraseFromParent();
+            }
+        }
+
+        /*
```

Comment:
In general, there shouldn't be any commented out code blocks. 
If having a detailed log while the tool runs, then it would be good to think about setting up a proper log level that allows to optionally enable a verbose output.

Reply:
Yes, eventually, we will need a logger. We do not have this at the moment, but that is something to be considered when introducing dependencies.